components/crud-web-apps/jupyter/backend/apps/common/routes/patch.py

```diff
@@ -0,0 +1,80 @@
+import datetime as dt
+
+from flask import request
+from werkzeug import exceptions
+
+from kubeflow.kubeflow.flask_rest_backend import api, decorators, logging
+
+from .. import status
+from . import bp
+
+log = logging.getLogger(__name__)
+
+STOP_ATTR = "stopped"
+ATTRIBUTES = set([STOP_ATTR])
+
+
+# Routes
+@bp.route(
+    "/api/namespaces/<namespace>/notebooks/<notebook>", methods=["PATCH"]
+)
+@decorators.request_is_json_type
+def patch_notebook(namespace, notebook):
+    request_body = request.get_json()
+    log.info("Got body: %s", request_body)
+
+    if request_body is None:
+        raise exceptions.BadRequest("Request doesn't have a body.")
+
+    # Ensure request has at least one valid command
+    if not any(attr in ATTRIBUTES for attr in request_body.keys()):
+        raise exceptions.BadRequest(
+            "Request body must include at least one supported key: %s"
+            % list(ATTRIBUTES)
+        )
+
+    # start/stop a notebook
+    if STOP_ATTR in request_body:
+        start_stop_notebook(namespace, notebook, request_body)
+
+    return api.success_response()
+
+
+# helper functions
+def start_stop_notebook(namespace, notebook, request_body):
```

Comment:
Why does this need to be a separate function?

The patch function is sufficient?

Reply:
I was thinking that If in the future we want to handle other cases in the PATCH request, then each one of them should be its own function and not have all the logic unfolded in the PATCH handler

Reply:
I was thinking that If in the future we want to handle other cases in the PATCH request, then each one of them should be its own function and not have all the logic unfolded in the PATCH handler